### PR TITLE
Fixed L-BFGS, included new code for BlockArray

### PR DESCRIPTION
**Pull request by lostella**
This PR is to fix L-BFGS: now it is implemented for both `Array` and `Tuple` objects at once.
Furthermore, the newly introduced `block.jl` defines more rigorously what `BlockArray` objects are (that is, `Array` or `Tuple` objects) and all the associated operations and convenient broadcasts. This will eventually serve as a replacement for `deep.jl`.

### CODE REVIEW

File: src/linearoperators/LBFGS.jl [hunk 1]
```diff
@@ -1,132 +1,107 @@
...
+	s_M::Array{T, 1}
+	y_M::Array{T, 1}
+	ys_M::Array{R, 1}
 	alphas::Array{R, 1}
 	H::R
 end
 
-# Constructors
-#default
-function LBFGS(T::Type, dim::NTuple{N,Int}, M::Int) where {N}
```
Comment (nantonel) [inline src/linearoperators/LBFGS.jl]: I would prefer to keep this constructor for consistency, this is the standard constructor for all operators:
`Operator(T::Type, dim::Tuple, args...)`

File: src/linearoperators/LBFGS.jl [hunk 2]
```diff
@@ -18,9 +18,9 @@ julia> d = L*grad; # compute new direction
 ```
 """
 
-mutable struct LBFGS{R, T <: BlockArray, M} <: LinearOperator
-	currmem::Integer
-	curridx::Integer
```
Comment (nantonel) [inline src/linearoperators/LBFGS.jl]: having Abstract Types in constructors creates Type Instability.

File: src/linearoperators/LBFGS.jl [hunk 3]
```diff
@@ -30,14 +30,32 @@ mutable struct LBFGS{R, T <: BlockArray, M} <: LinearOperator
 	H::R
 end
 
-function LBFGS(x::T, M::Integer) where {R, T <: BlockArray{R}}
```
Comment (nantonel) [inline src/linearoperators/LBFGS.jl]: this created problems as `R` for example when typing `LBFGS( (zeros(3),zeros(Complex{Float64})), 3)'. 
While `typeof( (zeros(3),zeros(Complex{Float64})) ) <: BlockArray{Float64} == true` apparently it couldn't be determined that `R` was `Float64`.

File: src/linearoperators/LBFGS.jl [hunk 4]
```diff
@@ -58,49 +76,49 @@ function update!(L::LBFGS{R, T, M}, x::T, x_prev::T, gradx::T, gradx_prev::T) wh
 		L.ys_M[L.curridx] = ys
 		blockcopy!(L.s_M[L.curridx], L.s)
 		blockcopy!(L.y_M[L.curridx], L.y)
-		yty = real(vecdot(L.y, L.y))
```
Comment (nantonel) [inline src/linearoperators/LBFGS.jl]: test passed also with `vecdot` but I guess it is more safe to use our `blockvecdot` no?

**Comment by nantonel** [comment]
definitively agree with that. changed that!

**Comment by lostella** [comment]
Ok! Let me propose some additional changes though, let's wait before merging.

**Comment by nantonel** [comment]
sure! actually why not waiting until the new solvers are done?

**Pull request merged**